### DISCUSSION TITLE
Is winget a kernel-mode program?

### DISCUSSION THREAD

**Discussion by Trenly** [microsoft/winget-cli#1821] [Q&A]
Had a really weird crash today while I was using winget to upgrade some packages. I believe it was due to an unhandled exception in a kernel-mode application which ended up in a memory access violation. I don't believe that any of the installers running would have had kernel-mode or interacted with kernel level drivers. This led me to wonder - is winget a kernel-mode program (running from administrative powershell)?

Just trying to narrow down the possibilities, since it did cause a small bit of corruption (repairable, thankfully) and I'm hoping to figure out what the cause was

**Comment by Masamune3210** [comment]
I'm not sure msix packages can even BE kernel mode

**Reply by Trenly** [reply]
I don't know if they can or can't be. But, I just want to rule out any possible sources, since an unhandled exception in ntoskernel seems pretty rare to me, and I feel as though it may have been caused by some other program trying to access kernel memory

**Reply by Masamune3210** [reply]
Maybe antivirus? They predominantly use kernel drivers so they can snoop everything and if they end up faulting they can cause a kernel-mode crash

**Reply by Trenly** [reply]
Don’t have any installed

**Comment by Masamune3210** [comment]
Windows Defender is force enabled on 10 and 11 if you don't have anything installed, unless you make an active effort to defeat it. Doubt its that by itself but maybe something interacting badly with it. Just throwing ideas out at this point

**Reply by Trenly** [reply]
Possibly, but I think that given the full circumstances and after looking at the system events and logs that it isn’t an antivirus or defender issue. It also doesn’t seem to be an audio or video driver issue. Since the only other things that were running at the time were winget upgrades and a few other miscellaneous applications, I believe those are the most likely culprits. I’m not looking for suggestions as to what *did* cause it, because I know that diagnosing bluescreen errors is extremely system dependent; Instead I’m trying to rule out what *didn’t* cause it, to narrow the scope of where I need to look before diving into application logs and memory dumps

**Reply by Masamune3210** [reply]
just remembered, if you open the memory dump from the bsod in a viewer it should be able to tell you the program that caused the fault, it might be able to tell you what driver was to blame

**Reply by Trenly** [reply]
> just remembered, if you open the memory dump from the bsod in a viewer it should be able to tell you the program that caused the fault, it might be able to tell you what driver was to blame

Yes, it does to an extent. I had looked into that as the first thing I did. However, it doesn’t quickly give any information on what application was using that driver at the time which caused the crash.

 In fact, it wasn’t a driver that bluescreened in this case. The crash came from `ntoskernel.exe`, specifically the bug check function, and specifically an access violation. Because of the way the kernel is written, this is the expected behavior when any application attempts to access memory not dedicated to it. In fact, this bug check and access violation behavior was talked about by David Plummer, one of the original developers on the NT OS. Because this is an intentional bluescreen called by the kernel itself, there has to be some other program trying to access the memory. Based on the memory address that the program tried to access and the specifc error code the kernel set when the bugcheck caught the access violation, I believe that the application trying to access them had to be running in kernel-mode. Of course this isn’t a guarantee, but again, I’m trying to narrow in on what it could have been before doing a deep dive.

Its entirely possible that the unhandled exception was just the kernel trying to access a peice of memory while it had a lock on it, but that seems unlikely to me, which is why I’m trying to start at the top and narrow down like a funnel
